### PR TITLE
Complete API spec implementation for flow_framework 

### PR DESCRIPTION
**Pull request by junweid62**
### Description
Part of Adding missing API specs https://github.com/opensearch-project/opensearch-api-specification/issues/168

API covered in this PR:

- /_plugins/_flow_framework/workflow/_steps:

-   /_plugins/_flow_framework/workflow/_search:

-   /_plugins/_flow_framework/workflow/state/_search:

-   /_plugins/_flow_framework/workflow/{workflow_id}/_deprovision:

-   /_plugins/_flow_framework/workflow/{workflow_id}/_provision:

-   /_plugins/_flow_framework/workflow/{workflow_id}/_status:

At this point, all FlowFramework APIs are covered

### Issues Resolved
https://github.com/opensearch-project/opensearch-api-specification/issues/508

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by junweid62** [comment]
@dbwiddis Just finished up the remaining API specs. I'd really appreciate it if you could check it out. Thanks!

File: CHANGELOG.md [hunk 1]
```diff
@@ -92,6 +92,7 @@ Inspired from [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 - Added request and response schemas for `/_cluster/decommission/awareness` ([#524](https://github.com/opensearch-project/opensearch-api-specification/pull/524))
 - Added `Bytes` component of type number ([#552](https://github.com/opensearch-project/opensearch-api-specification/pull/552))
 - Added `xy_shape` query ([#531](https://github.com/opensearch-project/opensearch-api-specification/pull/531))
+- Added `/_plugins/_flow_framework/_search` ,`state/_search`,`_provision`,`_deprovision`,`_steps`,`_status`([#508](https://github.com/opensearch-project/opensearch-api-specification/issues/508))
```
Comment (dbwiddis) [inline CHANGELOG.md:95]: Probably want to separate the `_search` from the first one, so it fits along with the rest.

File: spec/schemas/flow_framework.errors.yaml [hunk 2]
```diff
@@ -66,6 +66,35 @@ components:
               example: Workflow ID can not be null
             status:
               type: integer
+    DuplicateKeyError:
+      type: object
+      properties:
+        error:
+          type: string
+          description: Error message when a duplicate key is found in the request.
```
Comment (dbwiddis) [inline spec/schemas/flow_framework.errors.yaml:74]: Not sure "duplicate key" is self-descriptive here.  Is this a duplicate param, or a duplicate field in the body JSON?

File: spec/schemas/flow_framework.common.yaml [hunk 3]
```diff
@@ -6,6 +6,34 @@ info:
 paths: { }
 components:
   schemas:
+    AllowDelete:
+      description: A comma-separated list of resource IDs to be deprovisioned. Required if deleting resources of type index_name or pipeline_id.
```
Comment (dbwiddis) [inline spec/schemas/flow_framework.common.yaml:10]: Not sure if we should leave as is or make a bit more generic here.  I don't really like the specification of the specific type of resources. This specific list is correct as-of today but may be expanded and I think a change might get lost here.  Would prefer a more generic "required when deleting resources has a high potential for data loss" or similar.

**Review by dbwiddis — APPROVED** [review]
LGTM from a Flow Framework API perspective!  A few documentation nits/suggestions.

Did not review the technical details of the openapi spec.

File: tests/default/flow_framework/workflow.yaml [hunk 4]
```diff
@@ -49,11 +119,79 @@ chapters:
...
+    method: GET
+    parameters:
+      workflow_id: ${create_flow_framework.test_workflow_id}
+    response:
+      status: 200
+  - synopsis: Get workflow status with wrong workflow ID.
+    path: /_plugins/_flow_framework/workflow/{workflow_id}/_status
+    method: GET
+    parameters:
+      workflow_id: test
```
Comment (dblock) [inline tests/default/flow_framework/workflow.yaml:145]:
```suggestion
      workflow_id: invalid
```

Just a nit to be more explicit.

File: tests/default/flow_framework/workflow.yaml [hunk 5]
```diff
@@ -49,11 +119,79 @@ chapters:
...
-    id: delete_flow_framework
     path: /_plugins/_flow_framework/workflow/{workflow_id}
     method: DELETE
     parameters:
       workflow_id: ${create_flow_framework.test_workflow_id}
     response:
-      status: 200
\ No newline at end of file
+      status: 200
+  - synopsis: Delete workflow With Wrong ID.
```
Comment (dblock) [inline tests/default/flow_framework/workflow.yaml:189]:
```suggestion
  - synopsis: Delete workflow using an invalid ID.
```

**Review by dblock — APPROVED** [review]
Looks good. Will leave it unmerged to address various comments.

Consider breaking up the workflow test into workflow/privision.yaml, deprovision.yaml, etc. We are going to use these tests to provide samples for the documentation and users will be mostly looking to copy-paste examples to "privision a workflow" - it gets messy when there's too much going on and often setup may become unrelated (to deprovision a workflow you should have one privisioned in a prologue).

**Comment by dblock** [comment]
Looks like some legit test failures, might need to update the build ref in the CI.

```
        FAILED  Create workflow With Reprovision set true. (Operation "PUT /_plugins/_flow_framework/workflow" not found in the spec.)
        FAILED  Create workflow With Reprovision set true and UseCase Not null. (Operation "PUT /_plugins/_flow_framework/workflow" not found in the spec.)
        FAILED  Update workflow With Reprovision set true and UseCase Not null. (Operation "POST /_plugins/_flow_framework/workflow/{workflow_id}" not found in the spec.)
```

**Comment by dbwiddis** [comment]
> Looks like some legit test failures, might need to update the build ref in the CI.
> 
> ```
>         FAILED  Create workflow With Reprovision set true. (Operation "PUT /_plugins/_flow_framework/workflow" not found in the spec.)
>         FAILED  Create workflow With Reprovision set true and UseCase Not null. (Operation "PUT /_plugins/_flow_framework/workflow" not found in the spec.)
>         FAILED  Update workflow With Reprovision set true and UseCase Not null. (Operation "POST /_plugins/_flow_framework/workflow/{workflow_id}" not found in the spec.)
> ```

These seem reversed. POST ends at `/workflow`, PUT requires the workflow id.

File: tests/default/flow_framework/workflow.yaml (hunk 5 above)
Comment (dblock) [reply tests/default/flow_framework/workflow.yaml:189]: You don't like my suggestion? :)

File: spec/schemas/flow_framework.errors.yaml (hunk 2 above)
Comment (dblock) [reply spec/schemas/flow_framework.errors.yaml:74]: This looks correct to me, should match the code.

**Comment by junweid62** [comment]
@dblock Hey, I just pushed a new version. Please take a look when you get a chance!:)

**Comment by junweid62** [comment]
I was also wondering if there's a command that can remove all the empty lines?

**Comment by nhtruong** [comment]
@junweid62 most IDEs can do that for you "On Save". You should look up that feature for the IDE you're using.

File: spec/schemas/flow_framework.errors.yaml (hunk 2 above)
Comment (dbwiddis) [reply spec/schemas/flow_framework.errors.yaml:74]: For now as @dblock says it's ok to match the code. The error message should probably be "duplicate param" but not a big deal to leave it as is.  If you get the message you'll know what you did wrong.   [Whoever wrote that code](https://github.com/opensearch-project/flow-framework/pull/525) should have thought better. ;)

**Review by dblock — APPROVED** [review]

**Pull request merged**